KaitaiStream.cs

```diff
@@ -478,74 +488,150 @@ public static byte[] BytesTerminate(byte[] src, byte terminator, bool includeTer
         #region Byte array processing
 
         /// <summary>
-        /// Performs XOR processing with given data, XORing every byte of the input with a single value.
+        /// Perform XOR processing between given data and single-byte key.
+        /// WARNING: May return same byte array if key is zero.
         /// </summary>
-        /// <param name="value">The data toe process</param>
-        /// <param name="key">The key value to XOR with</param>
-        /// <returns>Processed data</returns>
-        public byte[] ProcessXor(byte[] value, int key)
+        /// <param name="data">The data to process, as byte array</param>
+        /// <param name="key">The key to XOR with, as integer</param>
+        public byte[] ProcessXor(byte[] data, int key)
         {
-            byte[] result = new byte[value.Length];
-            for (int i = 0; i < value.Length; i++)
+            if (key == 0)
+                return data;
+
+            int dl = data.Length;
+            byte[] result = new byte[dl];
+
+            for (int i = 0; i < dl; i++)
             {
-                result[i] = (byte)(value[i] ^ key);
+                result[i] = (byte)(data[i] ^ key);
             }
             return result;
         }
 
         /// <summary>
-        /// Performs XOR processing with given data, XORing every byte of the input with a key
-        /// array, repeating from the beginning of the key array if necessary
+        /// Perform XOR processing between given data and multiple-byte key.
+        /// WARNING: May return same byte array if key is zero.
         /// </summary>
-        /// <param name="value">The data toe process</param>
-        /// <param name="key">The key array to XOR with</param>
-        /// <returns>Processed data</returns>
-        public byte[] ProcessXor(byte[] value, byte[] key)
+        /// <param name="data">The data to process, as byte array</param>
+        /// <param name="key">The key to XOR with, as byte array</param>
+        public byte[] ProcessXor(byte[] data, byte[] key)
         {
-            int keyLen = key.Length;
-            byte[] result = new byte[value.Length];
-            for (int i = 0, j = 0; i < value.Length; i++, j = (j + 1) % keyLen)
+            if (key.Length == 1)
+                return ProcessXor(data, key[0]);
+            if (key.Length <= 64 && ByteArrayZero(key))
```

Comment:
You know [what I think about it](https://github.com/kaitai-io/kaitai_struct_python_runtime/pull/28).

Reply:
Optimisations are based on conjecture/assumption that keys are much shorter than data. That makes this check cheap in comparison.

Reply:
Also ByteArrayZero is short evaluation, it breaks loop on first non-zero.

Reply:
>Optimisations are based on conjecture/assumption that keys are much shorter than data. That makes this check cheap in comparison.

We don't need the assumption you are using ( `key.Length <= 64    <=>   Pr( key === zero bit string ) > key.Length/value.Length `) to be an inherent part of the runtime.